sound/soc/sof/intel/hda-dai.c

```diff
@@ -169,19 +169,24 @@ static int hda_link_pcm_prepare(struct snd_pcm_substream *substream,
 				struct snd_soc_dai *dai)
 {
 	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dpcm *dpcm;
 	struct snd_sof_dev *sdev =
 		snd_soc_component_get_drvdata(dai->component);
 	struct snd_sof_pcm *spcm;
 	int stream = substream->stream;
 
-	spcm = snd_sof_find_spcm_dai(sdev, rtd);
-	if (!spcm)
-		return -EINVAL;
+	for_each_dpcm_fe(rtd, stream, dpcm) {
+		spcm = snd_sof_find_spcm_dai(sdev, dpcm->fe);
```

Comment:
what if that BE is currently not connected to a FE due to mixer settings?

Reply:
@plbossart i saw your argument in Rander's PR about this too but I cant quite wrap it around my head. Are you talking about hostless pcm's?

Reply:
No, a front-end and a back-end connected by a switch which is off. Same as on Baytrail with the 'no BE error' without the mixers properly set-up.

Reply:
Google "Baytrail Audio Port: ASoC: no backend DAIs enabled for Baytrail Audio Port"

Reply:
@plbossart dpcm is certainly not my expertise. But let me take a stab at arguing that this case will be OK:
soc_dpcm_runtime_update() is called by DAPM mixer/mux changes. So when the switch is off, the runtime update will disconnect all BE dai links that were previously connected. 

I dont have any proof of this working as we dont have a pipeline with a mux to test it out. So this is purely what I understood from the dpcm flow.

Reply:
It seems we did not need to handle the dpcm cases here. We will only find some PCM that is connecting to the BE, then get the status from spcm, right?

Reply:
@xiulipan Yes, the purpose, I think, is to find whether the link dma used by the substream needs restore registers or not. However, spcm is used for fe. I think this is why the dpcm is used.

Reply:
> @plbossart dpcm is certainly not my expertise. But let me take a stab at arguing that this case will be OK:
> soc_dpcm_runtime_update() is called by DAPM mixer/mux changes. So when the switch is off, the runtime update will disconnect all BE dai links that were previously connected.
> 
> I dont have any proof of this working as we dont have a pipeline with a mux to test it out. So this is purely what I understood from the dpcm flow.

You could have 2 FEs that are connected to a BE in a mutually exclusive manner, or get mixed. So when a path is established, which FE is used completely depends on the configuration of the switches. If you have a mixer it's also unclear to me which FE you'd select.

We really need to build a topology with 2 BEs and 2 FEs, and a set of switches/mixer components to deal with the 2x2 connection. That's the only way to progress with those DPCM questions.